[DO NOT SQUASH] Add support for noalias atributes.

stefankoncarevic opened the pull request.

giuseros commented:
Nice one! It looks fine to me, but I think you should give it a round of `clang-format` because the formatting looks a bit odd :)
Also, could you write a description to help the reviewers understand the aim of the PR (maybe also pointing out the ticket you are working on?)

krzysz00 commented on external/llvm-project/mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp:
```diff
@@ -181,6 +186,30 @@ GPUFuncOpLowering::matchAndRewrite(gpu::GPUFuncOp gpuFuncOp, OpAdaptor adaptor,
+
+    if (memrefTy && argAttr.erase(LLVM::LLVMDialect::getNoAliasAttrName())) {
+      auto remapping = signatureConversion.getInputMapping(en.index());
+      for (size_t i = 0; i < remapping->size; i++)
```
Nit: LLVM house style calls for `++i` and, even more paranoidly, `(size_t i = 0, e = remapping->size; i < e; ++i)`

krzysz00 commented on external/llvm-project/mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp:
```diff
@@ -181,6 +186,30 @@ GPUFuncOpLowering::matchAndRewrite(gpu::GPUFuncOp gpuFuncOp, OpAdaptor adaptor,
+                                rewriter.getUnitAttr());
+        }
+    }
+    assert(argAttr.empty() && "Only noalias attributes support");
```
I'd rather not assert here and just let argument attributes fall through if unhandled. It's what the behavior was before and this could crash someone's working code.

krzysz00 commented on external/llvm-project/mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp:
```diff
@@ -181,6 +186,30 @@ GPUFuncOpLowering::matchAndRewrite(gpu::GPUFuncOp gpuFuncOp, OpAdaptor adaptor,
+    if (argAttr.empty())
+      continue;
+
+    if (memrefTy && argAttr.erase(LLVM::LLVMDialect::getNoAliasAttrName())) {
```
Let's generalize this a bit:
1. Stick a lambda together that handles copying an arbitrary pointer attribute (`noalias`, `readonly`, `nonnull`, `dereferencable_or_null(N)` and so on)
2. Instead of creating a new attribute, save the one you `erase()`d so you can add it to the pointer arguments

krzysz00 replied:
We don't actually have to do the general case, but it'll save everyone headache, I think.

krzysz00 commented on external/llvm-project/mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp:
```diff
@@ -181,6 +186,30 @@ GPUFuncOpLowering::matchAndRewrite(gpu::GPUFuncOp gpuFuncOp, OpAdaptor adaptor,
+        if (llvmFuncOp.getArgument(remapping->inputNo + i)
+                .getType()
+                .template isa<LLVM::LLVMPointerType>()) {
+          llvmFuncOp.setArgAttr(remapping->inputNo + i,
```
Correctness issue:
If `noalias` is present and the calling convention is not bare pointers (or, equivalently, if the remapping is of size > 1, currently), do **not** copy `noalias` and instead emit a warning. This is because memrefs, as you've probably noticed, have the allocation pointer and the aligned pointer, and those do, in fact, alias.

(In the future, we might want to add add an "aligned pointers only" scheme, but that's far future work we don't need)

krzysz00 commented on mlir/unittests/Dialect/Rock/TransformMapBuilderTests.cpp:
```diff
@@ -245,3 +250,28 @@ TEST_F(TMBuilderTest, GemmOut) {
   EXPECT_EQ(resDown, resUp);
 }
+
+TEST_F(TMBuilderTest, TestNoAliasAttributeForMemrefArgs) {
```
This is completely unneeded and not what you should've been changing.

krzysz00 reviewed (changes requested):
1. Correctness issue.
2. We might as well solve some of the general case while we're here
3. Add tests for the changes you just made to GPU function lowering:
You need to add a new file to `external/llvm-project/mlir/test/Conversion/GPUCommon/` . Use `lower-memory-space-attrs.mlir` in that directory as an example for how to set these tests up. Other tests in that directory or the surrounding ones may also be useful.

To be clear, I'm not upset, this is nonobvious stuff when you're new to the codebase.

krzysz00 commented:
Final note: this commit touches both our code and upstream. We'll want to send the upstream part to LLVM. To make that easy you'll need to:
1. Stick a `[DO NOT SQUASH]` in the title of this PR to warn everyone.
2. When merging it, use a merge commit
3. Most importantly, have the final work (you don't need to do this immediately, and you can just push commits here and then `rebase -i` your way to the correct form once you're approved) split into two commits: one containing only changes to `external/llvm-project/` and one containing all other changes.

The external commit should have a longer commit message (explaining, at a high level, what this is doing and why), and its first line will be something like `[mlir][GPU] Handle LLVM pointer attributes on memref arguments` or `[mlir][GPU] Handle llvm.noalias on memref arguments` depending on which change we actually make.

krzysz00 commented on external/llvm-project/mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp:
```diff
@@ -181,8 +186,46 @@ GPUFuncOpLowering::matchAndRewrite(gpu::GPUFuncOp gpuFuncOp, OpAdaptor adaptor,
+    auto remapping = signatureConversion.getInputMapping(en.index());
+    if (remapping->size > 1 &&
+        attrName == LLVM::LLVMDialect::getNoAliasAttrName()) {
+      mlir::emitWarning(
```
No need for a leading `mlir::`
Also, per upstream style, instead of `if {} else {}`ing, just `return;` after emitting the warning

krzysz00 commented on external/llvm-project/mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp:
```diff
@@ -181,8 +186,46 @@ GPUFuncOpLowering::matchAndRewrite(gpu::GPUFuncOp gpuFuncOp, OpAdaptor adaptor,
+      for (size_t i = 0, e = remapping->size; i < e; ++i) {
+        if (llvmFuncOp.getArgument(remapping->inputNo + i)
+                .getType()
+                .template isa<LLVM::LLVMPointerType>()) {
```
I don't think this needs a `template`

krzysz00 commented on external/llvm-project/mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp:
```diff
@@ -181,8 +186,46 @@ GPUFuncOpLowering::matchAndRewrite(gpu::GPUFuncOp gpuFuncOp, OpAdaptor adaptor,
+    if (!attr)
+      return;
+    auto remapping = signatureConversion.getInputMapping(en.index());
+    if (remapping->size > 1 &&
```
Since that's not an obvious condition, comment that this is a proxy for the bare pointer calling convention.

krzysz00 commented on external/llvm-project/mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp:
```diff
@@ -181,8 +186,46 @@ GPUFuncOpLowering::matchAndRewrite(gpu::GPUFuncOp gpuFuncOp, OpAdaptor adaptor,
+    copyPointerAttribute(LLVM::LLVMDialect::getNoAliasAttrName());
+    copyPointerAttribute(LLVM::LLVMDialect::getReadonlyAttrName());
+    copyPointerAttribute(LLVM::LLVMDialect::getNonNullAttrName());
+    copyPointerAttribute(LLVM::LLVMDialect::getDereferenceableOrNullAttrName());
```
I think straight-up `Dereferencable` exists as well, could go here?

krzysz00 reviewed (commented):
The main code looks file (aside from some style nitpicks), this just needs a test.

stefankoncarevic committed:
[mlir][GPU] Handle LLVM pointer attributes on memref arguments.

Handle pointer attributes (noalias, nonnull, readonly, writeonly,
dereferencable, dereferencable_or_null). "noalias" attribute is
ignore for non-bare pointer.

stefankoncarevic committed:
[mlir][GPU] Add noalias attribute on memref arguments

krzysz00 commented on external/llvm-project/mlir/test/Conversion/GPUCommon/memref-arg-attrs.mlir:
```diff
@@ -0,0 +1,50 @@
+
+// CHECK-LABEL:  llvm.func @test_func_dereferenceable_or_null
+// ROCDL-SAME:  !llvm.ptr {llvm.dereferenceable_or_null = 4 : i64}
+//  NVVM-SAME:  !llvm.ptr {llvm.dereferenceable_or_null = 4 : i64}
```
Nit: missing newline

krzysz00 commented on external/llvm-project/mlir/test/Conversion/GPUCommon/memref-arg-noalias-attrs.mlir:
```diff
@@ -0,0 +1,35 @@
+// RUN: mlir-opt %s -split-input-file -convert-gpu-to-rocdl='use-opaque-pointers=1 use-bare-ptr-memref-call-conv=1' | FileCheck %s --check-prefixes=CHECK,ROCDL
+// RUN: mlir-opt %s -split-input-file -convert-gpu-to-nvvm='use-opaque-pointers=1 use-bare-ptr-memref-call-conv=1' | FileCheck %s --check-prefixes=CHECK,NVVM
+// RUN: mlir-opt %s -split-input-file -convert-gpu-to-rocdl='use-opaque-pointers=1 use-bare-ptr-memref-call-conv=0' -o - 2>&1 | FileCheck %s --check-prefixes=CHECK-WARNING
```
This needs to be in a separate file `memref-arg-noalias-warning.mlir` that uses `-verify-diagnostics`

krzysz00 commented on external/llvm-project/mlir/test/Conversion/GPUCommon/memref-arg-noalias-attrs.mlir:
```diff
@@ -0,0 +1,35 @@
+// -----
+
+// CHECK-WARNING: Cannot copy noalias with non-bare pointers.
+gpu.module @kernel {
```
That is, this test in particular should be in its own file.

krzysz00 reviewed (commented):
Overall looks good, just need to shuffle around the diagnostic test

krzysz00 reviewed (approved):
Looks good to me, but could you add newlines to those test files.
